src/main/java/com/fakng/fakngagrgtr/scheduler/FakngScheduler.java

```diff
@@ -20,20 +20,26 @@ public class FakngScheduler {
 
     private final List<Parser> parsers;
     private final VacancyRepository vacancyRepository;
-    @Value("${agrgtr.pool-size:1}")
-    private int poolSize;
-    private ExecutorService executor;
+    @Value("${agrgtr.pool-size.primary:1}")
+    private int primaryPoolSize;
+
+    @Value("${agrgtr.pool-size.secondary:1}")
+    private int secondaryPoolSize;
+    private ExecutorService primaryExecutor;
+    private ExecutorService secondaryExecutor;
 
     @PostConstruct
     public void init() {
-        executor = Executors.newFixedThreadPool(poolSize);
+        primaryExecutor = Executors.newFixedThreadPool(primaryPoolSize);
+        secondaryExecutor = Executors.newFixedThreadPool(secondaryPoolSize);
     }
 
     @Scheduled(cron = "${agrgtr.cron:0 */12 * * * *}")
     public void scheduleAggregation() {
         parsers.forEach(parser -> CompletableFuture
-                .supplyAsync(parser::parse, executor)
-                .thenAccept(vacancyRepository::saveAll)
+                .supplyAsync(parser::parse, primaryExecutor)
+                .thenApply(vacancyRepository::saveAll)
+                .thenCompose(vacancies -> CompletableFuture.supplyAsync(() -> parser.parseSecondary(vacancies), secondaryExecutor))
```

Comment:
So when all vacancies are parsed and saved in DB, they are not fully constructed for some specific sites. For those, the second async parsing step is run. It's just about going checking a page of every vacancy and collecting the required data from there. 

One thing to think about here: is how to save only updated vacancies when this job completes. There's no need to update the entire batch if only 2 vacancies were added since the previous run.

Reply:
May be extract this functionality to API server and do this on user demand by hook. I did mention this in our talk